core/src/main/scala/fuse/package.scala

```diff
@@ -44,6 +44,124 @@ package object fuse {
   }
 
+  type FuseCompatibleFs = api.Filesystem {
+    type M[A] = Result[A]
+    type Path = String
+    type Name = String
+    type IO   = Array[Byte]
+  }
+
+  sealed trait Result[+A] {
+
+    def map[B](f: A => B): Result[B] =
+      this match {
+        case Success(a) => Success(f(a))
+        case x: Error   => x
+      }
+
+    def flatMap[B](f: A => Result[B]): Result[B] =
+      this match {
+        case Success(a) => f(a)
+        case x: Error   => x
+      }
+
+    def withFilter(f: A => Boolean): Result[A] =
+      this match {
+        case x @ Success(a) if f(a) => x
+        case _                      => InputOutputError
+      }
+
+    def orElseUse[AA >: A](z: => AA): Result[AA] =
+      this match {
+        case x @ Success(_) => x
+        case _: Error       => Success(z)
+      }
+
+    def orElse[AA >: A](z: => Result[AA]): Result[AA] =
+      this match {
+        case x @ Success(_) => x
+        case _: Error       => z
+      }
+
+    def whenSuccessful[AA >: A](z: => Result[AA]): Result[AA] =
+      this match {
+        case Success(_) => z
+        case x: Error   => x
+      }
+
+    def ensure(f: A =?> Unit): Result[A] =
+      this match {
+        case x @ Success(a) if f isDefinedAt a => x
+        case Success(_)                        => InputOutputError
+        case x: Error                          => x
+      }
+
+    def toInt()(implicit ev: A => Int): Int =
+      this match {
+        case Success(a) => a
+        case x: Error   => toErrorCode(x)
+      }
+  }
+
+  final case class Success[A](value: A) extends Result[A]
+  sealed trait Error extends Result[Nothing]
+
+  case object InputOutputError extends Error
+  case object AccessDenied     extends Error
+  case object TooBig           extends Error
+  case object NotEmpty         extends Error
+  case object NotImplemented   extends Error
+  case object NotValid         extends Error
+  case object DoesNotExist     extends Error
+  case object AlreadyExists    extends Error
+  case object NotSupported     extends Error
+
+  def toErrorCode[A]: Result[A] => Int = {
+    case DoesNotExist     => doesNotExist()
+    case NotValid         => isNotValid()
+    case NotImplemented   => notImplemented()
+    case NotEmpty         => -ENOTEMPTY
+    case TooBig           => -EFBIG
+    case AccessDenied     => -EACCES
+    case InputOutputError => -EIO
+    case _                => notImplemented()
+  }
+
+  implicit class NodeTypeOps(nodeType: api.attributes.NodeType) {
+
+    import api.attributes._
+    def asFuse = nodeType match {
+      case File => Node.File
+      case Dir  => Node.Dir
+      case Link => Node.Link
+    }
+  }
+
+  // This file exists to experiment with transforming parts of the file system, in this case the Path
+  implicit class Wrapped[FS <: api.Filesystem](val underlying: FS) {
+
+    def withPathAsString(
+      pathFromString: String => underlying.Path,
```

Comment:
Generic in what? As a point of interest, the whole thing is more generic than it needs to be. You never actually use `FS` anywhere. It's only needed because there's so much magic in implicit classes. If you break it into an implicit method and a class, then you can write it like
```scala
  implicit def Wrapped(u: api.Filesystem) = new Wrapped { val underlying: u.type = u }

  abstract class Wrapped {
    val underlying: api.Filesystem
    ...
  }
```